### PR TITLE
Fix for issue #21

### PR DESCRIPTION
**Pull request by andye2004**
Modified LinkedResourceMethodHandler to correctly update cached resources when setter method is called on proxy instance. Other small changes to support this feature included.

### CODE REVIEW

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    @Test
+    public void invokeSetsAndReturnsSameLinkedResource() throws InvocationTargetException, IllegalAccessException {
+        final Method setterMethod = findMethod(ResourceContent.class, "setSomething", String.class);
+        assertThat(setterMethod, is(notNullValue()));
+
+        final String setterValue = "STRING";
+        final Object setterResult = handler.invoke(resourceContent, setterMethod, null, new String[]{setterValue});
+
+        assertThat(setterResult, is(notNullValue()));
+        assertThat(setterResult, instanceOf(String.class));
```
Comment (andye2004) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: Just realised we only need the instanceOf assert, no need for the not null one too.

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        assertThat(setterValue, equalTo(setterResult));
```
Comment (andye2004) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: Just realised we only need the equalTo assert, no need for the other two.

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+        final Method getterMethod = findMethod(ResourceContent.class, "getSomething");
+        assertThat(getterMethod, is(notNullValue()));
+
+        final Object getterResult = handler.invoke(resourceContent, getterMethod, null, null);
+
+        assertThat(getterResult, is(notNullValue()));
+        assertThat(getterResult, instanceOf(String.class));
+        assertThat(getterResult, equalTo(setterResult));
```
Comment (andye2004) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: Ditto the above comment on the asserts

File: client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java [hunk 4]
```diff
@@ -1,59 +1,88 @@
...
+		super(resource);
 		this.restOperations = restOperations;
 		this.proxyFactory = proxyFactory;
 	}
-	
+
 	@Override
 	public boolean supports(Method method) {
-		return method.isAnnotationPresent(LinkedResource.class);
+		return method.isAnnotationPresent(LinkedResource.class) && (isSetter(method) || isGetter(method));
```
[the remark this answers is not in the document]
Comment (hdpe) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: We definitely don't want this because it will break #16 - linked resources must work with methods of any name.

[the remark this answers is not in the document]
Comment (andye2004) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: I hadn't read #16 but I did think the change a bit contentious, hence the reason for actually flagging it. Remove now and reverted the changes I had to make to test after adding it in.

**Comment by hdpe** [comment]
Thanks very much for raising this PR! It looks pretty solid. I'm quite busy at the moment so it may take me a while to properly review it, but please bear with me.

Thanks

**Comment by andye2004** [comment]
No worries Ryan, we're running the patch at the moment so no hold up for us anyway.

File: client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java [hunk 5]
```diff
@@ -1,59 +1,82 @@
...
+			linkedResourceResults.put(method.getName(), resolveLinkedResource(self, method, proceed, args));
 		}
-		
-		return linkedResourceResult;
+		return linkedResourceResults.get(method.getName());
+	}
+
+	private Object invokeSetter(Object self, Method method, Method proceed, Object[] args)
+	throws InvocationTargetException, IllegalAccessException {
+		super.invoke(self, method, proceed, args);
```
Comment (andye2004) [inline client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: Been mulling this one through, I'm not sure we actually need to call the underlying instance to set the link in the pojo. As long as the handler is called for the getter then it would return the updated link anyway. Just thinking out loud, be interested in your take on it.

Comment (hdpe) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: On a related point - if we merge the `GetterMethodHandler` and `SetterMethodHandler` similarly, there would then be no need for the abstract superclass now the `LinkedResourceMethodHandler` no longer needs it. I wonder if this could then become an `AbstractPropertyAwareMethodHandler` or similar and if this could contain the setter finding code etc., enabling you to DRY this a bit? What do you think?

Comment (andye2004) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: Hi Ryan sorry mate, been snowed under the last few hours. The ```LinkedResourceMethodHandler```, at least at the moment anyway, still requires access to the BeanInfo data as the links are mapped using the non setter method name. Based on your second comment this may not be an issue, assuming I understand your suggestion correctly.

Are you suggesting a ```PropertyAwareMethodHandler``` that is a delegate inside the ```LinkedResourceMethodHandler``` that is delegated to when the method being called is either a getter or setter and that the linked resource is actually stored in the proxied instance variable instead of the map? Then the map in the ```LinkedResourceMethodHandler``` only contains all other non-getter/setter links?

In this scenario the oder of handlers would be ```ResourceIdHandler```, ```LinkedResourceMethodHandler``` finally followed by the same instance of ```PropertyAwareMethodHandler``` passed to the ```LinkedResourceMethodHandler```?

Actually none of that sounds like what you're saying at all :)

Comment (andye2004) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: Actually I think I'm following now. An ```AbstractPropertyAwareMethodHandler``` that has all of the property aware methods and is then extended by a ```PropertyAwareMethodHandler``` and the ```LinkedResourceMethodHandler```?

Comment (hdpe) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: Yes exactly that - only maybe call the concrete impl `PropertyAwareMethodHandler` => `SimplePropertyMethodHandler` or some such - to show it's applicable to all getters and setters of "normal" (i.e. non-linked-resource annotated) properties.

Just a suggestion... what do you think?

Comment (andye2004) [reply client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: Yep, pretty much exactly what I have just updated to. Just tidying up the tests now and will update PR soon as complete.

File: client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java [hunk 6]
```diff
@@ -89,9 +108,9 @@ private Object resolveLinkedResource(Object self, Method method, Method proceed,
 		else {
 			collection.clear();
 		}
-		
-		for (Resource<F> resource : resources) {
-			collection.add(proxyFactory.create(resource, restOperations));
+
+		for (Resource<F> fResource : resources) {
```
Comment (andye2004) [inline client/src/main/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandler.java]: renamed to prevent clash with instance variable

File: client/src/test/java/uk/co/blackpepper/bowman/SimplePropertyAwareMethodHandlerTest.java [hunk 7]
```diff
@@ -66,4 +77,25 @@ public void supportsWithGetMethodWithVoidReturnTypeIsFalse() {
...
+
+	@Test
+	public void supportsWithSetMethodWithReturnTypeIsFalse() {
+		assertThat(handler.supports(findMethod(ResourceContent.class, "setThing3", String.class)), is(false));
+	}
+
+	@Test
+	public void supportsWithNonSetMethodIsFalse() {
+		assertThat(handler.supports(findMethod(ResourceContent.class, "notApplicable")), is(false));
+	}
```
Comment (andye2004) [inline client/src/test/java/uk/co/blackpepper/bowman/SimplePropertyAwareMethodHandlerTest.java]: Tests migrated from deleted SetterMethodHandlerTest

**Comment by hdpe** [comment]
Thanks again for all this. I think there's still a little bit of work to do here. I've pushed an integration test to new branch `issue-21-it` (I couldn't seem to push directly to your PR - perhaps because it's based on your master or you withheld permission?) which *I think* covered the use case you described in #21, but it's not passing. Please could you incorporate this test (or similar) into your PR, making it more applicable if need be?

Basically the `LinkedResourceMethodHandler` isn't getting called for setters, because the `@LinkedResource` annotation goes on the *getter*. So I think we need logic in `supports` more like:

1. If it's a setter, see if there's a `@LinkedResource` annotation on the corresponding getter; otherwise,
1. see if there's a `@LinkedResource` annotation on the method itself.

I think the test would then pass. What do you think?

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import static org.hamcrest.core.IsNull.notNullValue;
+import static org.junit.Assert.assertThat;
+import static org.springframework.util.ReflectionUtils.findMethod;
+
+public class LinkedResourceMethodHandlerTest {
+
+	private final ResourceContent resourceContent = new ResourceContent();
+	private final LinkedResourceMethodHandler handler = createHandler();
+
+	@Test
```
Comment (hdpe) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: I think these `supports*` tests are a bit misleading - the implementation doesn't care whether it's a getter that's annotated for instance, which might be confusing to the reader. I think your permutations here are:

1. `@LinkedResource` on method
1. `@LinkedResource` on setter's corresponding getter
1. No `@LinkedResource`

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	@Test
+	public void supportsNonAnnotatedGetterIsFalse() {
+		assertThat(handler.supports(findMethod(ResourceContent.class, "getSomethingElse")), is(false));
+	}
+
+	@Test
+	public void invokeSetsAndReturnsSameLinkedResource() throws InvocationTargetException, IllegalAccessException {
+		final Method setterMethod = findMethod(ResourceContent.class, "setSomething", String.class);
```
Comment (hdpe) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: Please could you make this test a bit simpler? Aim for one assertion per test if reasonably practicable. I don't think all those intermediate assertions add anything.

```java
handler.invoke(resourceContent, setterMethod, null, new String[]{"x"});

assertThat(handler.invoke(resourceContent, getterMethod, null, null), is("x"));
```

is nearly all you need.

File: client/src/test/java/uk/co/blackpepper/bowman/SimplePropertyAwareMethodHandlerTest.java [hunk 10]
```diff
@@ -8,8 +8,9 @@
 import static org.junit.Assert.assertThat;
 import static org.springframework.util.ReflectionUtils.findMethod;
 
-public class GetterMethodHandlerTest {
-	
+public class SimplePropertyAwareMethodHandlerTest {
+
+	@SuppressWarnings("ALL")
```
Comment (hdpe) [inline client/src/test/java/uk/co/blackpepper/bowman/SimplePropertyAwareMethodHandlerTest.java]: `"unused"` preferable here

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+		assertThat(getterResult, equalTo(setterResult));
+	}
+
+	private LinkedResourceMethodHandler createHandler() {
+		final JavassistClientProxyFactory proxyFactory = new JavassistClientProxyFactory();
+		return new LinkedResourceMethodHandler(new Resource<>(resourceContent),
+				new RestOperationsFactory(Configuration.builder().build(), proxyFactory).create(), proxyFactory);
+	}
+
+	@SuppressWarnings("ALL")
```
Comment (hdpe) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: prefer `@SuppressWarnings("unused")`

**Comment by andye2004** [comment]
@hdpe, For some strange reason I can't reply to your latest comments so I'll just respond here re your comment on the IT. In my original suggestion (and in the unit tests) I am assuming the setter actually is annotated with ```LinkedResource``` so _should_ have worked if the setter were annotated. That said I am happy to take up your suggestion re the check on the getter and have updated the PR accordingly.

All other suggestions / comments acred upon as requested.

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+		assertThat(handler.invoke(resourceContent, getterMethod, null, null), equalTo("X"));
+	}
+
+	private LinkedResourceMethodHandler createHandler() {
+		final JavassistClientProxyFactory proxyFactory = new JavassistClientProxyFactory();
+		return new LinkedResourceMethodHandler(new Resource<>(resourceContent),
+				new RestOperationsFactory(Configuration.builder().build(), proxyFactory).create(), proxyFactory);
+	}
+
+	@SuppressWarnings({"EmptyMethod", "SameReturnValue", "unused"})
```
Comment (hdpe) [inline client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: What *are* these warnings (EmptyMethod, SameReturnValue?) I've never encountered them before and can't find any documentation for them. What compiler or IDE are they targeting? Are they really necessary?

**Comment by hdpe** [comment]
Yes I don't think we want to have to annotate setters (to keep alignment with e.g. JPA - the JPA spec explicitly forbids this). Your new implementation looks good to me.

File: client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java (hunk 12 above)
Comment (andye2004) [reply client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: Those are IntelliJ static analysis warnings. If they don't show up on yours we can remove them.

Comment (hdpe) [reply client/src/test/java/uk/co/blackpepper/bowman/LinkedResourceMethodHandlerTest.java]: Yeah, get rid of them please in the couple of places they appear. Then we're good to go!

**Comment by hdpe** [comment]
Beautiful - thanks very much for raising this, for all your help, and for your support of this project!

**Comment by andye2004** [comment]
@hdpe, Pleasure Ryan, happy to help out! Especially given the usage we get out of the project. Happy to help if there is anything you need someone to look at in the future as well.

Can I just check when you think this will be released? Means I can keep an out and remove the dependency on our patched internal version.

**Comment by hdpe** [comment]
Released as 0.5.1.

Thanks
